### ISSUE TITLE
[pro] Can't use `cursor` or `endCursor` on a connection field which has a join in the relation

### ISSUE THREAD

**Issue by cheerfulstoic** [rmosolgo/graphql-ruby#1995]
I'm getting the error:

```
Failed to build GraphQL cursor, no attribute `cursor_0`
```

It trackes back to `graphql-pro-1.9.3/lib/graphql/pro/ordered_relation.rb:23`

I think it's because the `select` happening below doesn't get picked up by `ActiveRecord` when there are joins because `ActiveRecord` defines it's own columns and so only uses those to load attributes into the model.

Instead, couldn't it just access `node.attributes[v.name]` (or perhaps `node.send(v.name)` in case there's some sort of mapping)?  That would also mean that you wouldn't need to send an extra column's worth of data back from the DB

**Comment by rmosolgo** [comment]
Hey, thanks for reporting this! One trick to consider is that, when a relation is sorted by a SQL function, we also need to select the value of that function in order to generate the cursor. (The other option would be to reverse-engineer the function and implement it in ruby, using the attributes as input). 

But, maybe a hybrid solution would work in this case, reading from attributes when that is sufficient, but using extra `.select`s when necessary. 

Could you share an example relation that demonstrates this bug? (Maybe not real code, but just something that could inform my bug hunting! 🔍 )

**Comment by cheerfulstoic** [comment]
:+1: I didn't think about that (though I figured there was probably something I hadn't thought of ;) ).

I've made an example app to show the issue.  I was playing around with something else as well in there, but I'll give the info on how to reproduce this bug too.  The app:

https://github.com/cheerfulstoic/graphql_pro_bug_example_app

Of course you'll need to run the `bundle config` for the `graphql-pro` key.  In that app I can reproduce the issue with the following query:

```ruby
query {
  piratesCursorIssue(first: 1) {
    pageInfo {
      endCursor
      hasNextPage
    }
    edges { cursor node { id name age } }
  }
```

You can see that the implementation of `pirates_cursor_issue` is just `Pirate.joins(:ships).includes(:ships)`, though in my app I noticed it was actually occurring just with an `includes`, though I think that particular inclues might have been set up strangely to cause `ActiveRecord` to do a `JOIN` in SQL.  But I think a fix to this would fix the other.

**Comment by cheerfulstoic** [comment]
Oh, and I forgot to mention that you should do a `rake db:seed` in addition to the regular database setup so that the objects are there.

**Comment by rmosolgo** [comment]
Thanks for sharing that, I'll take a look soon!

**Comment by cheerfulstoic** [comment]
With regard to implementing it without both `joins` and `includes`... maybe..  For context: I have helper methods in my GraphQL schema classes which automatically do `includes` on associations which have been requested in the GraphQL query by looking at the `ast_node`.  In the field in question I'm using `joins` to filter out records according to it's association's details.  So at the moment I'm working around it with an ugly hack which ignores the specific association which is being automatically eager loaded.

But I suspect you might actually see this problem using just `includes` in certain cases.  Rails might have changed more recently, but in the past if you do an `includes` and then reference the associated table it makes one query rather than two and falls back to selecting the `t0_r0 ...` fields that you're seeing.  [The documentation](https://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-includes) for the `includes` method seems to imply that you need to call `references` to indicate that it needs to do that join.  So probably the magic condition parsing that they were doing previously is gone, but the use-case (people calling `references`) is likely still there.

So I don't know that I know what a solution might be, but I've subscribed to https://github.com/rails/rails/issues/34889 and I look forward to hearing what the Rails team has to say

**Comment by cheerfulstoic** [comment]
;( I've found another issue with the `cursor_0`.  I have a field with a method on the `UserType` class which is defined like:

```ruby
def field
  Model.recent(object)
end
```

That `recent` class method is defined as:

```ruby
def recent(user)
  Model.joins(:other_model).order(gmtDate: :desc).where(other_model_table: { owner: user }).distinct
end
```

So basically it's supposed to return a distinct list.  But it doesn't because the query is like this:

```sql
SELECT  DISTINCT model_table.*, other_model_table.gmtDate AS cursor_0, model_table.id AS cursor_1 FROM `model_table` INNER JOIN `other_model_table` ON `other_model_table`.`model_id` = `model_table`.`id` WHERE `other_model_table`.`owner` = 15 ORDER BY `model_table`.`gmtDate` DESC, model_table.id asc LIMIT 51
```

So because it's bringing `cursor_0` into the query the `DISTINCT` doesn't see a difference between the rows because the `gmtDate` is part of the result.

I still don't (yet) know of a solution, but adding a cursor to the queries is seeming more and more to me like a thing that is just going to bite people in various ways.

**Comment by rmosolgo** [comment]
A fix to the underlying Rails issue was just merged into Rails master 🍻

**Comment by cheerfulstoic** [comment]
🎉 I saw!  Since it's a bug, do you think it will come out in the next patch release?

**Comment by rmosolgo** [comment]
Yes, I would _expect_ so, I'll keep an eye on the release notes and follow up here!

**Comment by qrush** [comment]
Big +1 to getting a patch in for the next release. 🙌

**Comment by rmosolgo** [comment]
👋  I've just released graphql-pro 1.12.x which offers another approach to this issue. In GraphQL-Ruby 1.10.x (also just released), you can apply connection wrappers _inside the field resolver_ which gives you much more flexible control over which list objects get which connection behaviors. 

So, with graphql-pro 1.12.x (and graphql-ruby 1.10.x), you could enable the new stable relation connection (https://graphql-ruby.org/pagination/stable_relation_connections.html) and then _opt out_ on a field-by-field basis, falling back to index-based cursors when a stable cursor can't be created: 

```ruby
def items 
  items = Item.joins(...).select(...) # etc 
  # apply an index-based connection in this specific case:
  GraphQL::Pagination::ActiveRecordRelationConnection.new(items)
end 
```

Let's keep an eye on the upstream fix, too, but I thought I'd share this solution for now!

**Comment by rmosolgo** [comment]
It looks like the fix is due on the next release of Rails 6 (it's on the 6-0-stable changelog, https://github.com/rails/rails/blob/6-0-stable/activerecord/CHANGELOG.md), so I don't have any more work planned here. If this pops up again after you update the the next Rails version, please open a new issue!
